snapcraft/internal/build_providers/_multipass/_multipass_command.py

```diff
@@ -304,21 +317,85 @@ def umount(self, *, mount: str) -> None:
                 provider_name=self.provider_name, exit_code=process_error.returncode
             ) from process_error
 
-    def copy_files(self, *, source: str, destination: str) -> None:
-        """Passthrough for running multipass copy-files.
+    def push_file(self, *, source: IO, destination: str, bufsize: int = 1024) -> None:
+        """Passthrough for pushing a file through `multipass transfer`.
 
-        :param str source: the source file to copy, using syntax expected
-                           by multipass.
-        :param str destination: the destination of the copied file, using
-                                syntax expected by multipass.
+        :param IO source: a file-like object to read from
+        :param str destination: the destination of the copied file, using syntax
+                                expected by multipass
         """
-        cmd = [self.provider_cmd, "copy-files", source, destination]
-        try:
-            _run(cmd)
-        except subprocess.CalledProcessError as process_error:
-            raise errors.ProviderFileCopyError(
-                provider_name=self.provider_name, exit_code=process_error.returncode
-            ) from process_error
+        assert isinstance(source, io.IOBase)
+
+        # can't use std{in,out}=open(...) due to LP#1849753
+        data_to_read = True
+
+        p = _popen(
+            [self.provider_cmd, "transfer", "-", destination], stdin=subprocess.PIPE
+        )
+
+        while data_to_read:
+            read = source.read(bufsize)
+            p.stdin.write(read)
```

Comment:
This would result in an unnecessary `read()`, let me know if what I did now you think is ok, or whether you'd rather have the `read()` - it may be cheap?